File: 1501~2000/1877. Minimize Maximum Pair Sum in Array/main.py
"""
Source: https://rb.gy/3sa1i
Date: 2023/8/7
Skill:
Ref:
Runtime: 992 ms, faster than 90.06%
Memory Usage: 30.87 MB, less than 29.97%
Time complexity:
Space complexity:
Constraints:

"""

import math, sys
from typing import List, Optional
from collections import defaultdict, Counter, deque
from heapq import heapify, heappush, heappop, nsmallest
import heapq
import functools
from bisect import bisect_left, bisect_right


class Solution:
    def minPairSum(self, nums: List[int]) -> int:
        nums.sort()
        res, sz = 0, len(nums)
        for i in range(sz // 2):
            res = max(res, nums[i] + nums[sz - 1 - i])
        return res


if __name__ == "__main__":
    s = Solution()

File: 1501~2000/1877. Minimize Maximum Pair Sum in Array/main.cpp
/**
 * Source: ibit.ly/5FWNf
 * Date: 2023/11/17
 * Skill:
 * Runtime: 176 ms, faster than 87.58% of C++ online submissions
 * Memory Usage: 96.51 MB, less than 61.58% of C++ online submissions
 * Time complexity: O(n)
 * Space complexity: O(n)
 * Constraints:
 *
 */


#include <vector>
#include <iostream>
#include <algorithm>
#include <map>
#include <set>
#include <unordered_set>
#include <queue>
#include <numeric>
#include <tuple>
#include <stack>
#include <bitset>
#include <functional>
#include <list>


using namespace std;


class Solution {
public:
    int minPairSum(vector<int>& nums) {
        sort(begin(nums), end(nums));
        int res = INT32_MIN, n = nums.size();
        for (int i = 0; i < n / 2; i++) {
            res = max(res, nums[i] + nums[n - 1 - i]);
        }
        return res;
    }
};


int main() {
    Solution s;
}